### PR TITLE
feat: Associate role with appkeys

### PR DESCRIPTION
**Pull request by JigarJoshi**
## Describe your changes

- Associate role with app keys.
- Do not allow user to create appkey/user invitation with role higher than their own.

Note: This should not be merged prior to updating interfaces (cloud console, CLI) that accepts the role for app key creation.

## How best to test these changes
Added test.

### CODE REVIEW

**Review by efirs — DISMISSED** [review]

File: server/config/options.go [hunk 1]
```diff
@@ -101,6 +101,13 @@ type AuthConfig struct {
 	EnableErrorLog             bool                  `mapstructure:"enable_error_log" yaml:"enable_error_log" json:"enable_error_log"`
 }
 
+type ApiKeysConfig struct {
```
Comment (JigarJoshi) [inline server/config/options.go:104]: not related to this change but manual merge has moved it around.

File: server/middleware/authz.go [hunk 2]
```diff
@@ -32,10 +33,6 @@ const (
 )
 
 var (
-	// role names.
-	readOnlyRoleName     = "ro"
-	editorRoleName       = "e"
-	ownerRoleName        = "o"
```
Comment (JigarJoshi) [inline server/middleware/authz.go:38]: moved it to auth package to avoid circular dependencies.

File: server/middleware/authz.go [hunk 3]
```diff
@@ -450,6 +447,19 @@ func authorize(ctx context.Context) (err error) {
 	return nil
 }
 
+/*
+func isAuthorizedProject(reqMetadata *request.Metadata, accessToken *types.AccessToken) bool {
```
Comment (JigarJoshi) [inline server/middleware/authz.go:451]: keeping it here to enable back the project mismatch authz.

File: server/services/v1/auth/gotrue.go [hunk 4]
```diff
@@ -115,8 +122,15 @@ type GetUserResp struct {
 	AppMetaData *UserAppData `json:"app_metadata" db:"app_metadata"`
 }
 
-// returns currentSub, creationTime, error.
-func _createAppKey(ctx context.Context, clientId string, clientSecret string, g *gotrue, keyName string, keyDescription string, project string, keyType string) (string, int64, error) {
+type keyInfo struct {
```
Comment (JigarJoshi) [inline server/services/v1/auth/gotrue.go:125]: internal type to avoid long number of args.

File: server/services/v1/auth/gotrue.go [hunk 5]
```diff
@@ -157,20 +172,84 @@ func _createAppKey(ctx context.Context, clientId string, clientSecret string, g
...
 	return currentSub, creationTime, nil
 }
 
+func _getEmail(clientId string, appKeyType string, g *gotrue) string {
+	suffix := g.AuthConfig.Gotrue.UsernameSuffix
+	if appKeyType == AppKeyTypeApiKey {
+		suffix = g.AuthConfig.ApiKeys.EmailSuffix
+	}
+	return fmt.Sprintf("%s%s", clientId, suffix)
+}
```
Comment (JigarJoshi) [inline server/services/v1/auth/gotrue.go:187]: manual merge moved this method around.

File: test/v1/server/auth_test.go [hunk 6]
```diff
@@ -193,48 +193,127 @@ func TestGoTrueAuthProvider(t *testing.T) {
 		Boolean()
 	require.True(t, deletedResponse.Raw())
 }
+func TestApiKeyCrud(t *testing.T) {
```
Comment (JigarJoshi) [inline test/v1/server/auth_test.go:196]: manual merge moved it around.

**Review by efirs — APPROVED** [review]

**Pull request merged**